[mdoc] Fails to reflect Azure assemblies

Link to binaries:
https://1drv.ms/u/s!AuRndY7vE1CAh-YC0poPa2CkjhR_xw

```
2017-11-16T06:09:50.9670018Z Updating: Microsoft.ServiceBus.Messaging.Configuration.NetMessagingBindingExtensionElement
2017-11-16T06:09:59.2180450Z ##[error]mdoc: System.NullReferenceException: Object reference not set to an instance of an object.
   at Mono.Documentation.Updater.MsxdocDocumentationImporter.<>c.<ImportDocumentation>b__2_1(XmlNode n)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Mono.Documentation.Updater.MsxdocDocumentationImporter.ImportDocumentation(DocsNodeInfo info)
   at Mono.Documentation.MDocUpdater.MakeDocNode(DocsNodeInfo info, IEnumerable`1 setimporters)
   at Mono.Documentation.MDocUpdater.UpdateMember(DocsNodeInfo info, FrameworkTypeEntry typeEntry)
   at Mono.Documentation.MDocUpdater.DoUpdateType2(String message, XmlDocument basefile, TypeDefinition type, FrameworkTypeEntry typeEntry, String output, Boolean insertSince)
   at Mono.Documentation.MDocUpdater.DoUpdateType(TypeDefinition type, FrameworkTypeEntry typeEntry, String basepath, String dest)
   at Mono.Documentation.MDocUpdater.DoUpdateAssembly(AssemblySet assemblySet, AssemblyDefinition assembly, XmlElement index_types, String source, String dest, HashSet`1 goodfiles)
   at Mono.Documentation.MDocUpdater.DoUpdateAssemblies(String source, String dest)
   at Mono.Documentation.MDocUpdater.Run(IEnumerable`1 args)
   at Mono.Documentation.MDoc.Run(String[] args)
   at Mono.Documentation.MDoc.Main(String[] args)
See `mdoc help' for more information or use --debug to diagnose.
```

--- a/mdoc/Consts.cs
+++ b/mdoc/Consts.cs
@@ -3,7 +3,7 @@ namespace Mono.Documentation
 {
 	public static class Consts
 	{
-		public static string MonoVersion = "5.3.1.0";
+		public static string MonoVersion = "5.3.2.0";
 		public const string DocId = "DocId";
 		public const string VbNet = "VB.NET";
 		public const string DocIdLowCase = "docid";

--- a/mdoc/Mono.Documentation/MDocUpdater.cs
+++ b/mdoc/Mono.Documentation/MDocUpdater.cs
@@ -1398,7 +1398,7 @@ namespace Mono.Documentation
                             var sigvalue = sigFromXml.GetAttribute ("Value");
                             Func<FrameworkEntry, bool> findTypes = fx =>
                             {
-                                var tInstance = fx.Types.FirstOrDefault (t => t.Equals (typeEntry));
+                                var tInstance = fx.FindTypeEntry (typeEntry.Name);
                                 if (tInstance != null)
                                 {
                                     return tInstance.ContainsCSharpSig (sigvalue);

--- a/mdoc/Mono.Documentation/Updater/Frameworks/AssemblySet.cs
+++ b/mdoc/Mono.Documentation/Updater/Frameworks/AssemblySet.cs
@@ -7,69 +7,85 @@ using Mono.Cecil;
 
 namespace Mono.Documentation.Updater.Frameworks
 {
-	/// <summary>
-	/// Represents a set of assemblies that we want to document
-	/// </summary>
-	class AssemblySet : IDisposable
-	{
+    /// <summary>
+    /// Represents a set of assemblies that we want to document
+    /// </summary>
+    class AssemblySet : IDisposable
+    {
         readonly BaseAssemblyResolver resolver = new Frameworks.MDocResolver ();
         IAssemblyResolver cachedResolver;
-		HashSet<string> assemblyPaths = new HashSet<string> ();
-		HashSet<string> assemblySearchPaths = new HashSet<string> ();
-		HashSet<string> forwardedTypes = new HashSet<string> ();
+        HashSet<string> assemblyPaths = new HashSet<string> ();
+        Dictionary<string, bool> assemblyPathsMap = new Dictionary<string, bool> ();
+        HashSet<string> assemblySearchPaths = new HashSet<string> ();
+        HashSet<string> forwardedTypes = new HashSet<string> ();
         IEnumerable<string> importPaths;
-		public IEnumerable<DocumentationImporter> Importers { get; private set; }
+        public IEnumerable<DocumentationImporter> Importers { get; private set; }
 
-		public AssemblySet (IEnumerable<string> paths) : this ("Default", paths, new string[0], null) { }
+        public AssemblySet (IEnumerable<string> paths) : this ("Default", paths, new string[0], null) { }
 
         public AssemblySet (string name, IEnumerable<string> paths, IEnumerable<string> resolverSearchPaths, IEnumerable<string> imports)
-		{
+        {
             this.cachedResolver = new CachedResolver (this.resolver);
 
-			Name = name;
+            Name = name;
 
-			foreach (var path in paths)
-				assemblyPaths.Add (path);
+            foreach (var path in paths)
+            {
+                assemblyPaths.Add (path);
+                string pathName = Path.GetFileName (path);
+                if (!assemblyPathsMap.ContainsKey (pathName))
+                    assemblyPathsMap.Add (pathName, true);
+            }
 
-			// add default search paths
-			var assemblyDirectories = paths
-				.Where (p => p.Contains (Path.DirectorySeparatorChar))
-				.Select (p => Path.GetDirectoryName (p));
+            // add default search paths
+            var assemblyDirectories = paths
+                .Where (p => p.Contains (Path.DirectorySeparatorChar))
+                .Select (p => Path.GetDirectoryName (p));
 
-			foreach (var searchPath in resolverSearchPaths.Union (assemblyDirectories))
-				assemblySearchPaths.Add (searchPath);
+            foreach (var searchPath in resolverSearchPaths.Union (assemblyDirectories))
+                assemblySearchPaths.Add (searchPath);
 
-			char oppositeSeparator = Path.DirectorySeparatorChar == '/' ? '\\' : '/';
-			Func<string, string> sanitize = p =>
-				p.Replace (oppositeSeparator, Path.DirectorySeparatorChar);
+            char oppositeSeparator = Path.DirectorySeparatorChar == '/' ? '\\' : '/';
+            Func<string, string> sanitize = p =>
+                p.Replace (oppositeSeparator, Path.DirectorySeparatorChar);
 
-			foreach (var searchPath in assemblySearchPaths.Select (sanitize))
-				resolver.AddSearchDirectory (searchPath);
+            foreach (var searchPath in assemblySearchPaths.Select (sanitize))
+                resolver.AddSearchDirectory (searchPath);
 
-			this.importPaths = imports;
+            this.importPaths = imports;
             if (this.importPaths != null)
             {
-                this.Importers = this.importPaths.Select (p => MDocUpdater.Instance.GetImporter (p, supportsEcmaDoc: false));
+                this.Importers = this.importPaths.Select (p => MDocUpdater.Instance.GetImporter (p, supportsEcmaDoc: false)).ToArray ();
             }
             else
                 this.Importers = new DocumentationImporter[0];
-		}
+        }
 
-		public string Name { get; private set; }
+        public string Name { get; private set; }
 
-		public IEnumerable<AssemblyDefinition> Assemblies { get { return this.LoadAllAssemblies ().Where(a => a != null); } }
-		public IEnumerable<string> AssemblyPaths { get { return this.assemblyPaths; } }
+        IEnumerable<AssemblyDefinition> assemblies;
+        public IEnumerable<AssemblyDefinition> Assemblies
+        {
+            get
+            {
+                if (this.assemblies == null)
+                    this.assemblies = this.LoadAllAssemblies ().Where (a => a != null).ToArray ();
+
+                return this.assemblies;
+            }
+        }
+        public IEnumerable<string> AssemblyPaths { get { return this.assemblyPaths; } }
 
         /// <summary>Adds all subdirectories to the search directories for the resolver to look in.</summary>
-        public void RecurseSearchDirectories() 
+        public void RecurseSearchDirectories ()
         {
             var directories = resolver
                 .GetSearchDirectories ()
-                .Select(d => new DirectoryInfo (d))
+                .Select (d => new DirectoryInfo (d))
                 .Where (d => d.Exists)
-                .Select(d => d.FullName)
+                .Select (d => d.FullName)
                 .Distinct ()
-                .ToDictionary(d => d, d => d);
+                .ToDictionary (d => d, d => d);
 
             var subdirs = directories.Keys
                 .SelectMany (d => Directory.GetDirectories (d, ".", SearchOption.AllDirectories))
@@ -79,21 +95,25 @@ namespace Mono.Documentation.Updater.Frameworks
                 resolver.AddSearchDirectory (dir);
         }
 
-		/// <returns><c>true</c>, if in set was contained in the set of assemblies, <c>false</c> otherwise.</returns>
-		/// <param name="name">An assembly file name</param>
-		public bool Contains (string name)
-		{
-			return assemblyPaths.Any (p => Path.GetFileName (p) == name);
-		}
+        /// <returns><c>true</c>, if in set was contained in the set of assemblies, <c>false</c> otherwise.</returns>
+        /// <param name="name">An assembly file name</param>
+        public bool Contains (string name)
+        {
+            return assemblyPathsMap.ContainsKey (name);//assemblyPaths.Any (p => Path.GetFileName (p) == name);
+        }
 
-		/// <summary>Tells whether an already enumerated AssemblyDefinition, contains the type.</summary>
-		/// <param name="name">Type name</param>
-		public bool ContainsForwardedType (string name)
-		{
-			return forwardedTypes.Contains (name);
-		}
+        /// <summary>Tells whether an already enumerated AssemblyDefinition, contains the type.</summary>
+        /// <param name="name">Type name</param>
+        public bool ContainsForwardedType (string name)
+        {
+            return forwardedTypes.Contains (name);
+        }
 
-        public void Dispose () => cachedResolver.Dispose ();
+        public void Dispose () 
+        {
+            this.assemblies = null;
+            cachedResolver.Dispose();
+        }
 
 		public override string ToString ()
 		{

--- a/mdoc/Mono.Documentation/Updater/Frameworks/FrameworkEntry.cs
+++ b/mdoc/Mono.Documentation/Updater/Frameworks/FrameworkEntry.cs
@@ -24,6 +24,13 @@ namespace Mono.Documentation.Updater.Frameworks
 		public IEnumerable<DocumentationImporter> Importers { get; set; }
 
 		public ISet<FrameworkTypeEntry> Types { get { return this.types; } }
+        Dictionary<string, FrameworkTypeEntry> typeMap = new Dictionary<string, FrameworkTypeEntry> ();
+
+        public FrameworkTypeEntry FindTypeEntry (string name) {
+            FrameworkTypeEntry entry;
+            typeMap.TryGetValue (name, out entry);
+            return entry;
+        }
 
 		public IEnumerable<FrameworkEntry> Frameworks { get { return this.allframeworks; } }
 
@@ -31,11 +38,14 @@ namespace Mono.Documentation.Updater.Frameworks
 
 		public virtual FrameworkTypeEntry ProcessType (TypeDefinition type)
 		{
-			var entry = types.FirstOrDefault (t => t.Name.Equals (type.FullName, StringComparison.Ordinal));
-			if (entry == null) {
+            FrameworkTypeEntry entry;
+
+            if (!typeMap.TryGetValue (type.FullName, out entry)) {
 				var docid = DocCommentId.GetDocCommentId (type);
 				entry = new FrameworkTypeEntry (this) { Id = docid, Name = type.FullName, Namespace = type.Namespace };
 				types.Add (entry);
+
+                typeMap.Add (entry.Name, entry);
 			}
 			return entry;
 		}

--- a/mdoc/Mono.Documentation/Updater/Frameworks/FrameworkTypeEntry.cs
+++ b/mdoc/Mono.Documentation/Updater/Frameworks/FrameworkTypeEntry.cs
@@ -9,6 +9,7 @@ namespace Mono.Documentation.Updater.Frameworks
 	{
 		SortedSet<string> members = new SortedSet<string> ();
 		SortedSet<string> memberscsharpsig = new SortedSet<string> ();
+        Dictionary<string, bool> sigMap = new Dictionary<string, bool> ();
 
 		ILFullMemberFormatter formatter = new ILFullMemberFormatter ();
 
@@ -44,14 +45,17 @@ namespace Mono.Documentation.Updater.Frameworks
 
 			// this is for lookup purposes
 			try {
-				memberscsharpsig.Add(formatter.GetDeclaration(member));
+                var sig = formatter.GetDeclaration (member);
+				memberscsharpsig.Add(sig);
+                if (!sigMap.ContainsKey (sig))
+                    sigMap.Add (sig, true);
 			}
 			catch {}
 		}
 
 		public bool ContainsCSharpSig (string sig)
 		{
-			return memberscsharpsig.Contains (sig);
+			return sigMap.ContainsKey (sig);
 		}
 
 		public override string ToString () => $"{this.Name} in {this.fx.Name}";

--- a/mdoc/Mono.Documentation/Updater/MsxdocDocumentationImporter.cs
+++ b/mdoc/Mono.Documentation/Updater/MsxdocDocumentationImporter.cs
@@ -130,7 +130,7 @@ namespace Mono.Documentation.Updater
                                 .Select (n => new
                                 {
                                     Xml = n.OuterXml,
-                                    Overwrite = n.Attributes["overwrite"]
+                                    Overwrite = n.Attributes != null ? n.Attributes["overwrite"] : null
                                 });
                             string sourceXml = child.OuterXml;
 

--- a/mdoc/mdoc.nuspec
+++ b/mdoc/mdoc.nuspec
@@ -2,7 +2,7 @@
 <package >
   <metadata>
     <id>mdoc</id>
-    <version>5.3.1.0</version>
+    <version>5.3.2.0</version>
     <title>mdoc</title>
     <authors>Joel Martinez</authors>
     <owners>Xamarin</owners>

